xla/service/gpu/runtime/copy_thunk.h

```diff
@@ -55,22 +59,70 @@ class DeviceToDeviceCopyThunk : public Thunk {
   const uint64_t mem_size_;
 };
 
-class DeviceToHostCopyThunk : public DeviceToDeviceCopyThunk {
- public:
-  DeviceToHostCopyThunk(ThunkInfo thunk_info,
-                        const BufferAllocation::Slice& source_buffer,
-                        const BufferAllocation::Slice& destination_buffer,
-                        uint64_t mem_size);
-  absl::Status ExecuteOnStream(const ExecuteParams& params) override;
+//===----------------------------------------------------------------------===//
+// CopyAsyncEvents
+//===----------------------------------------------------------------------===//
+class CopyAsyncEvents {
+public:
+  // Add a new copy-start completion event.
+  absl::Status Emplace(se::StreamExecutor *executor,
+                       const HloInstruction *instr, se::Event &&event);
+
+  // Retrieve a completion event started by copy-start instruction
+  // `instr`, and remove the event from the collection.
+
+  absl::StatusOr<se::Event> Extract(se::StreamExecutor *executor,
+                                    const HloInstruction *instr);
+
+private:
+  using Key = std::pair<se::StreamExecutor *, const HloInstruction *>;
+  absl::Mutex mutex_;
+  absl::flat_hash_map<Key, se::Event> events_ ABSL_GUARDED_BY(mutex_);
 };
 
-class HostToDeviceCopyThunk : public DeviceToDeviceCopyThunk {
- public:
-  HostToDeviceCopyThunk(ThunkInfo thunk_info,
-                        const BufferAllocation::Slice& source_buffer,
-                        const BufferAllocation::Slice& destination_buffer,
-                        uint64_t mem_size);
-  absl::Status ExecuteOnStream(const ExecuteParams& params) override;
+//===----------------------------------------------------------------------===//
+// DeviceHostCopyThunk
+//===----------------------------------------------------------------------===//
+// The memcpy between a host and a device
+enum class CopyDirection { kHostToDevice = 0, kDeviceToHost = 1 };
+
+// A thunk that copies data from a device buffer to a host buffer.
+class DeviceHostCopyThunk : public DeviceToDeviceCopyThunk {
```

Comment:
`DeviceHostCopy` should not really inherit from `DeviceToDeviceCopy`, it's too confusing, I suggest to add `CopyThunk` and inherit d2d and d2h thunks from it

Reply:
Added CopyThunk and inherit d2h and h2d thunks from it.